File: 백준/Platinum/22949. 회전 미로 탐색/회전 미로 탐색.py
from collections import deque, namedtuple

def printBoard(board):
    for y in range(size):
        for x in range(size):
            if isinstance(board[y][x],str):
                print(f"{board[y][x]:3s}",end=' ')
            elif isinstance(board[y][x], int):
                print(f"{board[y][x]:3d}", end=' ')
        print()

def rotateAll(board):

    ret = [['.' for _ in range(size)] for _ in range(size)]

    half = size // 2
    for sy in range(0,size, half):
        for sx in range(0, size, half):
            ey,ex = sy + half - 1, sx + half - 1
            for y in range(sy, ey+1):
                for x in range(sx, ex+1):
                    y1, x1 = y - sy, x - sx
                    y2, x2 = x1, half - y1 - 1
                    ny, nx = y2 + sy, x2 + sx
                    ret[ny][nx] = board[y][x]

    return ret

dy = [0,1,-1,0,0]
dx = [0,0,0,1,-1]

def OOB(y,x):
    return y< 0 or y >= size or x< 0 or x >= size




def getDivision(y, x):
    if OOB(y,x):
        return -1
    return (y // 4) * 4 + (x // 4)


def getRotatePos(y, x):
    baseY, baseX = (y // 4) * 4, (x // 4) * 4

    y %= 4
    x %= 4

    return baseY + x, baseX + 3 - y


if __name__ == '__main__':
    N = int(input())
    size = 4 * N
    board = [[['.' for _ in range(size)] for _ in range(size)] for _ in range(4)]
    INF = float('inf')
    visited = [[[False for _ in range(size)] for _ in range(size)] for _ in range(4)]

    sy = sx = ey = ex = -1
    for y in range(size):
        line = input()
        for x in range(size):
            board[0][y][x] = line[x]
            if board[0][y][x] == 'S':
                sy,sx = y,x

            tmpY = y
            tmpX = x

            for rDir in range(1,4):
                tmpY, tmpX = getRotatePos(tmpY, tmpX)
                board[rDir][tmpY][tmpX] = board[0][y][x]

    # for rDir in range(1, 4):
    #     board[rDir] = rotateAll(board[rDir - 1])

    # printBoard(board[2])

    Tuple = namedtuple('Tuple', ['y','x','bDir', 'dist'])
    visited[0][sy][sx] = True
    q = deque([Tuple(sy,sx,0, 0)])

    while q:
        cur = q.popleft()

        if board[cur.bDir][cur.y][cur.x] == 'E':
            print(cur.dist)
            exit()

        cDiv = getDivision(cur.y, cur.x)
        for moveDir in range(5):
            ny = cur.y + dy[moveDir]
            nx = cur.x + dx[moveDir]
            nDiv = getDivision(ny, nx)

            if nDiv == -1:
                continue

            nxtbDir = (cur.bDir + 1) % 4 if cDiv == nDiv else 1

            ny, nx = getRotatePos(ny, nx)
            # print(nxtbDir, ny,nx)
            if board[nxtbDir][ny][nx] == '#' or visited[nxtbDir][ny][nx]:
                continue

            q.append(Tuple(ny,nx,nxtbDir, cur.dist+1))
            visited[nxtbDir][ny][nx] = True

    print(-1)

File: 백준/Platinum/22949. 회전 미로 탐색/회전 미로 탐색.java
import java.util.*;
import java.io.*;

public class Main {
 
    public static void main(String[] args) throws IOException{
    	BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    	N = Integer.parseInt(br.readLine());
    	size = 4 * N;
    	
 
    	init(br);
    	
    	dist[0][start.y][start.x] = 0;
    	Queue<Tuple> q = new ArrayDeque<>();
    	q.add(new Tuple(start.y, start.x, 0));
    	
    	
    	
    	while(!q.isEmpty()) {
    		Tuple cur = q.poll();
    		
    		if(board[cur.d][cur.y][cur.x] == 'E') {
    			System.out.println(dist[cur.d][cur.y][cur.x]);
    			System.exit(0);
    		}
    		
    		int div = getDivision(cur.y, cur.x);
    		
    		for(int moveDir = 0; moveDir < 5; moveDir++) {
    			int ny = cur.y + dy[moveDir];
    			int nx = cur.x + dx[moveDir];
    			
    			if(OOB(ny,nx))
    				continue;
    			
    			int nDiv = getDivision(ny,nx);
    			if(nDiv == -1)
    				continue;
    			
    			int nd = (div == nDiv) ? (cur.d + 1) % 4 : 1;
    			Pair nxt = rotate(ny,nx);
    			if(board[nd][nxt.y][nxt.x] == '#' || dist[nd][nxt.y][nxt.x] != INF)
    				continue;
    			dist[nd][nxt.y][nxt.x] = dist[cur.d][cur.y][cur.x] + 1;
    			q.add(new Tuple(nxt.y,nxt.x, nd));
    		}
    	}
    	System.out.println(-1);
    	
    }
    
    static Pair rotate(int y, int x) {
		int sy = (y / 4) * 4;
		int sx = (x / 4) * 4;
		
		int y1 = y - sy;
		int x1 = x - sx;
		
		int y2 = x1;
		int x2 = 3 - y1;
		
		int y3 = y2 + sy;
		int x3 = x2 + sx;
		
		return new Pair(y3,x3);
    }
    
    static int dy[] = {0,-1,1,0,0};
    static int dx[] = {0,0,0,-1,1};
    
    static int getDivision(int y,int x) {
    	if(OOB(y,x))
    		return -1;
    	
    	return (y / 4) * N + (x / 4) + 1;
    }
    
    
    static boolean OOB(int y,int x) {
    	return y<0 || y>=size || x<0 || x >= size;
    }
    
    
    static void init(BufferedReader br) throws IOException {
    	board = new char[4][size][size];
    	
    	dist = new int[4][size][size];
    	
       	for(int y=0;y<size;y++) {
    		String line = br.readLine();
    		for(int x= 0; x<size;x++) {
    			board[0][y][x] = line.charAt(x);
    			
    			if(board[0][y][x] == 'S') {
    				start = new Pair(y,x);
    			}
    			
    			int curY = y;
    			int curX = x;
    			dist[0][y][x] = INF;
    			for(int i = 1; i<4;i++) {
    				dist[i][y][x] = INF;
    				
    				Pair nxt = rotate(curY, curX);
    				
    				board[i][nxt.y][nxt.x]= board[i-1][curY][curX];
    				curY = nxt.y;
    				curX = nxt.x;
    				
    			}
    		}
    	}
    }
    static Pair start;
    static void printBoard(char board[][]) {
    	for(int y=0; y<size;y++) {
    		for(int x=0; x<size;x++) {
    			System.out.printf("%3s", board[y][x]);
    		}
    		System.out.println();
    	}
    }
    
    static int N;
    static int size;
    static int sy,sx;
    static int INF = Integer.MAX_VALUE / 2;
    static char board[][][];
    static int dist[][][];
    static class Pair{
    	int y,x;
    	public Pair(int y,int x) {
    		this.y = y;
    		this.x = x;
    	}
    }
    
    static class Tuple{
    	int y,x,d;
    	public Tuple(int y,int x, int d) {
    		this.y = y;
    		this.x = x;
    		this.d = d;
    	}
    }
}